_compdemos/python_logging_demo.md

```diff
@@ -0,0 +1,236 @@
+[here](https://docs.python.org/3.8/library/logging.html#logging.LogRecord).  Some of the attributes I find myself using
+quite often are `%(lineno)d` (line number), `%(asctime)s` (timestamp), `%(filename)`, `(funcName)s`, and `%(module)s`.  
+There are also some handy formatting attributes for reporting what thread or process a message is coming from if you're 
+working in a concurrent environment.
+
+When creating a Formatter object, it simply must be added to a Handler class instance using the `.setFormatter()` function,
+as seen in the earlier example.  By attaching different `Formatter` objects to different handlers, you can provide different levels
+of detail to different outputs.  For instance maybe I don't want the module, function, and file name cluttering up
+my console, so I could attach a formatter to it which doesn't have those attributes.  But I want my file handler to have as
+much information as possible so I can better reconstruct events when doing retrospectives (example below).
+
+```python
+import logging
+logger = logging.getLogger('new_logger')
+logger.setLevel(logging.DEBUG)
+file_formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s - line %(lineno)s - module %(module)s - function %(funcName)s - file %(fileName)s')
+file_handler = logging.FileHandler(filename='verbose_logger.log')
+file_handler.setFormatter(file_formatter)
+
+slim_formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s - line %(lineno)s')
+slim_console_handler = logging.StreamHandler()
+slim_console_handler.setFormatter(slim_formatter)
+logger.addHandler(slim_console_handler)
+logger.addHandler(file_handler)
+```
+
+### Filters 
+`Filters` are attached to `Handlers` and provided filtering of messages from below a specified point in the logger hierarchy 
+of which the `Logger` that the attached `Handler` is a part of.  We didn't go through logger hierarchies too much, but say
+we have a logger hierarchy like:
+```  
+- base_logger  
+- - module_logger  
+- - debug_logger
+```
+where the parent-child relationships of our loggers can be described as `base_logger.module_logger` and `base_logger.debug_logger`. 
+A `Filter` object can be attached to a `Logger` instance in order to prevent it from propagating messages from specific parts of the 
+logger hierarchy. `Filter` objects can also be attached to specific handlers as well, in order for more nuanced control of which loggers 
+are emitted to a particular output location.  
+For instance, if we were to want to have messages from our `module_logger` propagate to the `base_logger`, but did not want any messages 
+from `debug_logger` propagating up to the `base_logger` and its handlers, a `Filter` object could be attached to the `base_logger`
+specifying that it should ignore any messages recorded by `debug_logger`   
+
+### Log levels  
+Both `Logger` and `Handler` instances have an attribute called `.level`, which determines the severity of messages the 
```

Comment:
his presently isn't shown above in the code snippet: the filter is being applied at the logger level-- perhaps a second code snippet adding two FileHandlers: infoLogs and errorLogs with different message levels specified?

Reply:
A simple example of two different fileHandlers with different log levels might be a great way to explain log levels.  I'll consider adding a code snippet for that.  Thanks!